### ISSUE TITLE
mixed 401/200 with basic auth - zypper refresh - password ignored?

### ISSUE THREAD

**Issue by phillxnet** [openSUSE/zypper#359]
Apologies up front if this is not a bug, but I'm getting mixed Apache log entries re authentication when using basic http auth:
Repo setup. And on occasion a 'zypper refresh' post repo add requests user (pre-filled) and password entry before proceeding.

## Repo added via:
```
/usr/bin/zypper --non-interactive addrepo --refresh http://User:pass@updates.rockstor.com:8999/rockstor-stable/leap/15.2 Rockstor-Stable5
```
This completes as expected and a "zypper refresh", where no user/pass prompt appears, results, in order with Apache logs:
## Unexpected Auth failing calls (repomd.xml):
```
...81 - - [14/Oct/2020:12:23:30 +0000] "HEAD /rockstor-stable/leap/15.2/repodata/repomd.xml HTTP/1.1" 401 - "-" "ZYpp 17.24.2 (curl 7.66.0) openSUSE-Leap-15.2-x86_64"
...81 - - [14/Oct/2020:12:23:30 +0000] "HEAD /rockstor-stable/leap/15.2/repodata/repomd.xml HTTP/1.1" 401 - "-" "ZYpp 17.24.2 (curl 7.66.0) openSUSE-Leap-15.2-x86_64"
```
## Expected Auth success on same file target (repomd.xml):
```
...81 - User [14/Oct/2020:12:23:30 +0000] "HEAD /rockstor-stable/leap/15.2/repodata/repomd.xml HTTP/1.1" 200 - "-" "ZYpp 17.24.2 (curl 7.66.0) openSUSE-Leap-15.2-x86_64"
```
##  Auth fail - file doesn't exist in my createrepo_c created rpm repo.
```
...81 - - [14/Oct/2020:12:23:30 +0000] "GET /rockstor-stable/leap/15.2/media.1/media HTTP/1.1" 401 1258 "-" "ZYpp 17.24.2 (curl 7.66.0) openSUSE-Leap-15.2-x86_64"
...81 - - [14/Oct/2020:12:23:30 +0000] "GET /rockstor-stable/leap/15.2/media.1/media HTTP/1.1" 401 1258 "-" "ZYpp 17.24.2 (curl 7.66.0) openSUSE-Leap-15.2-x86_64"
```
## Expected not found for above missing media.1/media file:
```
..81 - - [14/Oct/2020:12:23:30 +0000] "GET /rockstor-stable/leap/15.2/media.1/media HTTP/1.1" 404 992 "-" "ZYpp 17.24.2 (curl 7.66.0) openSUSE-Leap-15.2-x86_64"
```
## Auth success on signing file "repomd.xml.asc"
```
...81 - User [14/Oct/2020:12:23:30 +0000] "GET /rockstor-stable/leap/15.2/repodata/repomd.xml.asc HTTP/1.1" 200 516 "-" "ZYpp 17.24.2 (curl 7.66.0) openSUSE-Leap-15.2-x86_64"
```
## Expected missing for repomd.xml.key
```
...81 - - [14/Oct/2020:12:23:30 +0000] "GET /rockstor-stable/leap/15.2/repodata/repomd.xml.key HTTP/1.1" 404 992 "-" "ZYpp 17.24.2 (curl 7.66.0) openSUSE-Leap-15.2-x86_64"
```
## Auth success on repomd.xml again + repodata file GET:
```
...81 - User [14/Oct/2020:12:23:30 +0000] "GET /rockstor-stable/leap/15.2/repodata/repomd.xml HTTP/1.1" 200 3082 "-" "ZYpp 17.24.2 (curl 7.66.0) openSUSE-Leap-15.2-x86_64"
...81 - User [14/Oct/2020:12:23:30 +0000] "GET /rockstor-stable/leap/15.2/repodata/aadb9ca63ef04003eb89b3d1e0b5399f5442058232023
8b27296de16dbdbaa36-primary.xml.gz HTTP/1.1" 200 1476 "-" "ZYpp 17.24.2 (curl 7.66.0) openSUSE-Leap-15.2-x86_64"
```
## Two Matching error.log Password Mismatch entries:
```
[Wed Oct 14 12:23:30.319671 2020] [auth_basic:error] [pid 24870] [client ...81:37302] AH01617: user User: authentication failure for "/rockstor-stable/leap/15.2/repodata/repomd.xml": Password Mismatch
[Wed Oct 14 12:23:30.478985 2020] [auth_basic:error] [pid 25481] [client ...81:37306] AH01617: user User: authentication failure for "/rockstor-stable/leap/15.2/media.1/media": Password Mismatch
```

Post repo addition there is knowledge of the repos contents:
```
zypper se -s rockstor
Loading repository data...
Reading installed packages...

S  | Name                              | Type       | Version          | Arch   | Repository
---+-----------------------------------+------------+------------------+--------+-----------------------------------
i+ | rockstor                          | package    | 4.0.0-0          | x86_64 | (System Packages)
v  | rockstor                          | package    | 4.0.3-0          | x86_64 | Rockstor-Stable5
i  | systemd-presets-branding-rockstor | package    | 12.2-lp152.196.1 | noarch | home_rockstor_branches_Base_System
   | systemd-presets-branding-rockstor | srcpackage | 12.2-lp152.196.1 | noarch | home_rockstor_branches_Base_System

```
## Repo's package update can succeed but can also on occasions fail:
Here again we get both failed and succeeded authentication: 2 401 followed by a 200.
```
...81 - - [14/Oct/2020:13:12:53 +0000] "GET /rockstor-stable/leap/15.2/rockstor-4.0.3-0.x86_64.rpm HTTP/1.1" 401 1272 "-" "ZYpp 17.24.2 (curl 7.66.0) openSUSE-Leap-15.2-x86_64"
...81 - - [14/Oct/2020:13:12:53 +0000] "GET /rockstor-stable/leap/15.2/rockstor-4.0.3-0.x86_64.rpm HTTP/1.1" 401 1272 "-" "ZYpp 17.24.2 (curl 7.66.0) openSUSE-Leap-15.2-x86_64"
...81 - User [14/Oct/2020:13:12:53 +0000] "GET /rockstor-stable/leap/15.2/rockstor-4.0.3-0.x86_64.rpm HTTP/1.1" 200 14100004 "-" "ZYpp 17.24.2 (curl 7.66.0) openSUSE-Leap-15.2-x86_64"
```
To my only partially informed 'eye' I see both unauthenticated access and correct authenticated access.

I'm strongly suspect I'm doing-it-wrong but I'm also very interested in ensuring we do what we can to report findings when we come across them. Again I must appologies if this is entirely misplaced but it does look odd to have both successful and un-sucessful basic http auth result from the above single repo add.

Also note that I don't see the entered 'passwored' in the autocreated /etc/zypp/repos.d/Rockstor-Stable5.repo file:
```
[Rockstor-Stable5]
enabled=1
autorefresh=1
baseurl=http://User@updates.rockstor.com:8999/rockstor-stable/leap/15.2
```
I would have expected to see in the above "User:pass@" instead of only "User@" but have assumed thus far that this is as a result of a mechanism to avoid plain text passwored or the like. Note also that the output of the addrepo command also does not reproduce the password component.
```
leap/15.2 Rockstor-Stable5
Adding repository 'Rockstor-Stable5' ......................................................................................................................................[done]
Repository 'Rockstor-Stable5' successfully added

URI         : http://User@updates.rockstor.com:8999/rockstor-stable/leap/15.2
Enabled     : Yes
GPG Check   : Yes
Autorefresh : Yes
Priority    : 99 (default priority)

Repository priorities in effect:                                                                                                                 (See 'zypper lr -P' for details)
      97 (raised priority)  :  2 repositories
      99 (default priority) :  3 repositories
     105 (lowered priority) :  1 repository
```
Again I'd assumed the 'missing' passwored in the addrepo command output was to not reproduce the password in plain text, so that it doesn't appear in log entries also. Which I've not seen it do incidentally.

But the overall behaviour of both auth success and failure, seemingly intertwined, and my experience of both success and failure in being able to update from this repo also suggests I'm either doing it wrong or have missed a vital step here somewhere. Or this is a bug, hence my report.

Also note that we do have rpm GPG key auth for this repo pre-added:
```
rpm -qa --qf '%{VERSION}-%{RELEASE} %{SUMMARY}\n' gpg-pubkey*
5f043187-5ed2a099 gpg(The Rockstor Project (Rockstor Development) <support@rockstor.com>)
307e3d54-5aaa90a5 gpg(SuSE Package Signing Key <build@suse.de>)
3dbdc284-53674dd4 gpg(openSUSE Project Signing Key <opensuse@opensuse.org>)
17280ddf-5e82f96b gpg(network OBS Project <network@build.opensuse.org>)
b91e1e8b-5be66216 gpg(shells OBS Project <shells@build.opensuse.org>)
39db7c82-5847eb1f gpg(SuSE Package Signing Key <build@suse.de>)
fbe3abff-5c518e74 gpg(home:rockstor OBS Project <home:rockstor@build.opensuse.org>)
```
These are added as part of the image build detailed below.

So if you could indicate if the auth failures we are seeing are a know issue and if they may have been fixed already, or if I'm just doing it wrong then that would much appreciated.

Note that we get far more sane log entries if I manually add the password to the /etc/zypp/repos.d/Rockstor-Stable5.repo file.

My suspicion her is that there are 2 processes associated with the zypper/package function where one is using the entered password credentials (but from where I don't see) and the other is not, or is just using the username with no associated pass as per the contents of zypper repo definition file.

## Upstream Zypper version used:
OS version: "Built on openSUSE"  Leap 15.2 JeOS 'alike' instance.
built via kiwi-ng config https://github.com/rockstor/rockstor-installer
using all upstream packages var our 2 rpms pre-isntalled.
```
zypper info zypper | grep 'Version\|Status'
Version        : 1.14.38-lp152.2.6.1
Status         : up-to-date
```

Thanks again for your attention and the work that goes into zypper.
Our stable http authed repo is our attempt to achieve our own sustainable open source development via a curated updates channel along side our non authed testing channel. Configured identically, bar not auth, and tested against zypper for months now successfully.

**Comment by phillxnet** [comment]
As additional information I have also become aware of the zyppre / libzypp aspect. So from the instance:
```
zypper info libzypp | grep 'Version\|Status'
Version        : 17.24.2-lp152.2.10.1
Status         : up-to-date
```
And from: https://build.opensuse.org/package/view_file/zypp:Head/libzypp/libzypp.changes?expand=1
I see in that libzypp changelog for a  just before this one:

> -------------------------------------------------------------------
> Thu Jul 16 18:04:18 CEST 2020 - ma@suse.de
> 
> - Fix bsc#1174011 auth=basic ignored in some cases (bsc#1174011)
>   Proactively send credentials if the URL specifes '?auth=basic'
>   and a username.
> - ZYPP_MEDIA_CURL_DEBUG: Strip credentials in header log (bsc#1174011)
> - version 17.24.1 (22)

Does this point more toward the 'your doing it wrong' conclusion I'd entertained above: I have no "?auth=basic" element to my url used in the above detailed "zypper addrepo ..."?

**Comment by mlandres** [comment]
@phillxnet sorry for the late answer.

A client does not itself decide to send authentication data. The server requests it by returning `401 Unauthorized` and setting a `WWW-Authenticate:` header. The header tells which authentication methods the server accepts. Then the client repeats the request using one of the supported authentication methods. This is part of the common HTTP authentication workflow. You don't send any auth data in advance without knowing whether the server will accept the method.

bsc#1174011 is not a bug, but a workaround for a specific problem. The server missbehaves and blocks access after a couple of '401' responses. Forcing the backend to support '?auth=basic' as the only authentication method, we send available credentials with every request rather than waiting for a '401'.

**Comment by mlandres** [comment]
@phillxnet  regarding the password in the URL: It is stripped by addrepo and stored in `~\.zypp/credentials.cat`. This is done because the `/etc/zypp/repos.d/*.repo` files are usually world readable. `zypper lr -u` executed by any user would hide the password, but `grep ^baseurl /etc/zypp/repos.d/*` would still reveal it.

If you explicitly add the password again you may get a behavior similar to adding '?auth=basic'. But that's implementation dependent.

**Comment by phillxnet** [comment]
@mlandres Thanks for the response and explanation here. Much appreciated. That does clear some things up.

In the end I had assumed I was doing it wrong or not going along with the more common way of setting thing up. From memory I think I proved the ~\.zypp/credentials.cat route but that failed for our exact implementation as the python env within which we required the zypper auth to work failed to fetch root's env: a known issue apparently that was deemed 'wont fix', referenced in my issue and below. In the end we moved to populating our own credentials file, as kiwi-ng does, and specifying that to zypper explicitly and all then working as intended within our python environment.

In case my final approach, workaround or otherwise, is of interest the pull request to the project concerned was:
https://github.com/rockstor/rockstor-core/pull/2228
issue:
https://github.com/rockstor/rockstor-core/issues/2227
which in turn sites the env 'wont fix' mentioned above, referenced here for convenience:
https://tickets.puppetlabs.com/browse/PUP-1124

Given I now suspect I was not appreciating the env difficulty and our need to move more with the times, as we did in the end, I'll close this issue as 'initially confused' and 'still a little confused but now working via indicated approach' essentially copied from kiwi-ng: doing the same from within a python environment. I believe SLES repo subscriptions also use this same explicit credentials file approach as well with their use of authenticated repositories along with zypper.

Thanks again for you explanation and apologies for leaving this issue hanging.
